File: java/sourceafis/src/sourceafis/extraction/minutiae/MinutiaCloudRemover.java
/**
 * @author Veaceslav Dubenco
 * @since 18.10.2012
 */
package sourceafis.extraction.minutiae;

import java.util.Iterator;

import sourceafis.general.Calc;
import sourceafis.general.DetailLogger;
import sourceafis.meta.Parameter;
import sourceafis.templates.Minutia;
import sourceafis.templates.TemplateBuilder;

/**
 * 
 */
public final class MinutiaCloudRemover {
	@Parameter(upper = 300)
	public int NeighborhoodRadius = 20;
	@Parameter(upper = 30)
	public int MaxNeighbors = 4;

	public DetailLogger.Hook Logger = DetailLogger.off;

	public void Filter(TemplateBuilder template) {
		int radiusSq = Calc.Sq(NeighborhoodRadius);
		Iterator<Minutia> iter = template.minutiae.iterator();
		while (iter.hasNext()) {
			Minutia minutia = iter.next();
			int count = 0;
			for (Minutia neighbor : template.minutiae) {
				if (Calc.DistanceSq(neighbor.Position, minutia.Position) <= radiusSq) {
					count++;
				}
			}
			if (count - 1 > MaxNeighbors) {
				iter.remove();
			}
		}
		Logger.log(template);
	}
}

File: SourceAFIS/Extraction/Minutiae/MinutiaCloudRemover.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SourceAFIS.General;
using SourceAFIS.Meta;
using SourceAFIS.Templates;

namespace SourceAFIS.Extraction.Minutiae
{
    public sealed class MinutiaCloudRemover
    {
        [Parameter(Upper = 300)]
        public int NeighborhoodRadius = 20;
        [Parameter(Upper = 30)]
        public int MaxNeighbors = 4;

        public DetailLogger.Hook Logger = DetailLogger.Null;

        public void Filter(FingerprintTemplate template)
        {
            var radiusSq = Calc.Sq(NeighborhoodRadius);
            template.Minutiae = template.Minutiae.Except(
                (from minutia in template.Minutiae
                 where template.Minutiae.Count(neighbor => Calc.DistanceSq(neighbor.Position, minutia.Position) <= radiusSq) - 1 > MaxNeighbors
                 select minutia).ToList()).ToList();
            Logger.Log(template);
        }
    }
}
